bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java

```diff
@@ -0,0 +1,81 @@
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.automation.module.mqtt.handler;
+
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.automation.Condition;
+import org.eclipse.smarthome.automation.handler.BaseModuleHandler;
+import org.eclipse.smarthome.automation.handler.ConditionHandler;
+import org.eclipse.smarthome.binding.mqtt.discovery.MQTTTopicDiscoveryParticipant;
+import org.eclipse.smarthome.binding.mqtt.discovery.MQTTTopicDiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.io.transport.mqtt.MqttBrokerConnection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Observes a MQTT topic. Checks if the current topic value matches a configured one.
+ *
+ * @author David Graeff - Initial Contribution
+ *
+ */
+@NonNullByDefault
+public class PublishedMessageConditionHandler extends BaseModuleHandler<Condition>
+        implements ConditionHandler, MQTTTopicDiscoveryParticipant {
+
+    private final Logger logger = LoggerFactory.getLogger(PublishedMessageConditionHandler.class);
+
+    public static final String MODULE_TYPE_ID = "mqtt.PublishedMessage";
+    public static final String CFG_TOPIC = "topic";
+    public static final String CFG_PAYLOAD = "payload";
+
+    private final MQTTTopicDiscoveryService discoveryService;
+    private final String expectedPayload;
+    private String receivedPayload = "";
+
+    public PublishedMessageConditionHandler(Condition module, MQTTTopicDiscoveryService discoveryService) {
+        super(module);
+        this.discoveryService = discoveryService;
+        expectedPayload = (String) module.getConfiguration().get(CFG_PAYLOAD);
+        final String topic = (String) module.getConfiguration().get(CFG_TOPIC);
+        discoveryService.subscribe(this, topic);
+    }
+
+    @Override
+    public void dispose() {
+        discoveryService.unsubscribe(this);
+        super.dispose();
+    }
+
+    @Override
+    public void receivedMessage(ThingUID thingUID, MqttBrokerConnection connection, @NonNull String topic,
+            byte @NonNull [] payload) {
+        receivedPayload = new String(payload);
+        logger.trace("Payload changed: {}", receivedPayload);
+    }
+
+    @Override
+    public void topicVanished(ThingUID thingUID, MqttBrokerConnection connection, @NonNull String topic) {
+        receivedPayload = "";
+    }
+
+    @NonNullByDefault({}) // TODO Until the ConditionHandler interface is fixed
+    @Override
+    public boolean isSatisfied(Map<String, Object> context) {
+        return expectedPayload.equals(receivedPayload);
```

Comment:
Because MQTT allows retained topic values (which translates into a state for a topic), the condition also makes perfect sense to me. The description might need to be changed to reflect the use case. The condition of course does not make sense for non-retained topics.

The trigger should fire on every new published message on a given topic, there are people in the community that require this behavior. But I do see the point. Another trigger could be introduced that triggers only on a specific message on a given topic.

Reply:
> Because MQTT allows retained topic values

But this implementation does not actively "request" those retained values from the broker or does it? What would be the use case where one would need them instead of relying on an item being filled with a value and using the existing item condition?

> Another trigger could be introduced that triggers only on a specific message on a given topic.

No! The existing MQTT trigger should simply have an optional parameter - this is exactly how it is done in the ItemStateTrigger, so we should be in line with that.

Reply:
Agreed. So an optional configuration for a trigger message will be added.

The implementation does not actively subscribe to retained values, that is correct. It's not possible actually. You can filter messages later on, if you want to, because messages have the retained flag set by the broker. That is not possible in the current MQTTBrokerConnection implementation though (you are usually not interested if a topic value is retained or not).

Reply:
Sounds like a good reason then to remove the condition - at least for now.

Reply:
What about this use-case:

**WHEN** *mqtt_topic_updated*("sensors/window1") with *mqtt_message*("open")
**IF** *mqtt_topic*("alarmpanel/state") *equals* *mqtt_message*("ARMED")
**THEN** ...

MQTT topics are not necessarily bound to an item.

Reply:
> " not necessarily"

Correct. But in your example, the "alarmpanel state" probably would make a very good item as well. And my big fear is people will start mixing stuff as in your example and end up with parts of their system state representation in items and other parts in the mqtt broker's retained events. That definitely should not be our design goal. I see MQTT as an "extension to the ESH event bus" and the item registry is meant to be the place, where event values are persisted (just as it has been [documented 8 years ago ;-)](https://github.com/openhab/openhab1-addons/wiki#item-repository)).